### PR TITLE
[GEP-20] Zone pinning for shoots on multi-zonal seeds 📌 

### PR DESCRIPTION
**Pull request by timuthy**
**How to categorize this PR?**

/area control-plane
/area high-availability
/kind enhancement

**What this PR does / why we need it**:
This PR adds the zone pinning requirement discussed by GEP-20, [ref](https://github.com/gardener/gardener/blob/master/docs/proposals/20-ha-control-planes.md#zone-pinning).

In essence it does the following:
- Gardenlet labels namespaces which belong to `non-HA` or `single-zonal` shoots if it's responsible for a `multi-zonal` seed (seed which has worker across availability zones).
- A `pod-zone-affinity` webhook in GRM adds a zone-aware affinity term for such control-plane pods so that those will be scheduled into a single zone only.
- After `etcd` was deployed successfully, it's assumed that at least one stateful pod was scheduled. Gardenlet then extracts the zone information from that pod/node.
- From that moment on, pods' affinity configurations is enhanced by a `nodeAffinity` to "pin" pods to a specific zone. It is especially important to schedule pods to the original zone when shoots are being woken up from hibernation.

**Which issue(s) this PR fixes**:
Fixes partially #6529

**Release note**:

```breaking operator
Gardener has been being prepared for more shoot HA use-cases and thus some assumption about currently running landscapes are required: If you use a `multi-zonal` labelled seed and scheduled non-HA shoots onto it, this release of Gardener will potentially cause scheduling conflicts to the control-plane pods as it will try to locate all pods into a single zone only. Pods that can't be re-scheduled (mainly because of volume dependencies) will remain in `Pending` state.
```
```feature operator
Gardener is prepared to run non-HA and single-zonal shoots on multi-zonal seeds. In such a setup, control-plane pods of the mentioned shoots are scheduled into a single availability zone only to avoid any extra cross zonal traffic that would usually involve higher latency and cost. **PLEASE NOTE**: The `StorageClass` in seeds used for control-plane components must have `volumeBindingMode: WaitForFirstConsumer` to let the zone-pinning work properly.
```

### CODE REVIEW

**Comment by timuthy** [comment]
/hold
I plan to improve the affinity terms after having a second look. With the current approach we rely on the fact, that any pod with a volume is created and scheduled first. This becomes especially important when a shoot is woken up from hibernation. If a stateless pod is scheduled first to any zone, then the remaining stateful pods are supposed to be scheduled to exactly this zone. Unfortunately, there is no guarantee that this stateless pod will run in very same zone that was used before the cluster was hibernated.

This PR will be transferred into draft mode as long as I'm working on the improvement.

**Comment by timuthy** [comment]
Update: The PR is ready for review now.

File: pkg/operation/botanist/namespaces.go [hunk 1]
```diff
@@ -74,6 +76,17 @@ func (b *Botanist) DeploySeedNamespace(ctx context.Context) error {
...
 
+		// Label namespace to pin all control-plane pods of a shoot cluster to one zone
+		// if the seed has workers across different availability zones.
+		if _, ok := b.Seed.GetInfo().Labels[v1beta1constants.LabelSeedMultiZonal]; ok {
+			haAnnotationValues, haIsConfigured := b.Shoot.GetInfo().Annotations[v1beta1constants.ShootAlphaControlPlaneHighAvailability]
+
+			if (haAnnotationValues == v1beta1constants.ShootAlphaControlPlaneHighAvailabilitySingleZone && gardenletfeatures.FeatureGate.Enabled(features.HAControlPlanes)) ||
+				!haIsConfigured {
+				metav1.SetMetaDataLabel(&namespace.ObjectMeta, v1beta1constants.ShootZonePinning, "")
+			}
```
Comment (oliver-goetz) [inline pkg/operation/botanist/namespaces.go]: GEP 20 mentions migrating non-HA shoots to HA shoots, so I'm wondering if there is a need for removing the zone pinning annotation again.

File: pkg/gardenlet/controller/shoot/shoot_control_reconcile.go [hunk 2]
```diff
@@ -281,6 +281,11 @@ func (r *shootReconciler) runReconcileShootFlow(ctx context.Context, o *operatio
 			Fn:           flow.TaskFn(botanist.WaitUntilEtcdsReady).SkipIf(o.Shoot.HibernationEnabled),
 			Dependencies: flow.NewTaskIDs(deployETCD),
 		})
+		_ = g.Add(flow.Task{
+			Name:         "Adding zone information to Shoot namespace",
+			Fn:           flow.TaskFn(botanist.AddZoneInformationToSeedNamespace).SkipIf(o.Shoot.HibernationEnabled),
+			Dependencies: flow.NewTaskIDs(waitUntilEtcdReady),
+		})
```
Comment (oliver-goetz) [inline pkg/gardenlet/controller/shoot/shoot_control_reconcile.go:288]: Shouldn't the subsequent `deployControlPlane` task wait until the zone information are added to the namespace to ensure, that they are not scheduled to the wrong zone?

File: pkg/resourcemanager/webhook/podzoneaffinity/handler.go [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+func removeConflictingPodAntiAffinityTerms(pod *corev1.Pod, affinityTerm corev1.PodAffinityTerm) {
+	if pod.Spec.Affinity == nil || pod.Spec.Affinity.PodAntiAffinity == nil {
+		return
+	}
+
+	// Remove pod anti-affinity rules with zone topology key.
+	remainingAntiAffinityTerms := make([]corev1.PodAffinityTerm, 0, len(pod.Spec.Affinity.PodAntiAffinity.RequiredDuringSchedulingIgnoredDuringExecution))
+	for _, term := range pod.Spec.Affinity.PodAntiAffinity.RequiredDuringSchedulingIgnoredDuringExecution {
+		if equality.Semantic.DeepEqual(term, affinityTerm) {
+			continue
```
Comment (oliver-goetz) [inline pkg/resourcemanager/webhook/podzoneaffinity/handler.go]: Is it worth mentioning that `PodAntiAffinity` rules in an annotation or so for debugging purposes?

File: pkg/resourcemanager/webhook/podzoneaffinity/handler.go [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+
+	for _, term := range pod.Spec.Affinity.PodAffinity.RequiredDuringSchedulingIgnoredDuringExecution {
+		if equality.Semantic.DeepEqual(term, affinityTerm) {
+			zoneTermExisting = true
+		}
+
+		// If there is another affinity configured on zones, we assume that this will be conflicting.
+		if term.TopologyKey == corev1.LabelTopologyZone && !zoneTermExisting {
+			continue
+		}
```
Comment (oliver-goetz) [inline pkg/resourcemanager/webhook/podzoneaffinity/handler.go:151]: The same here. How about saving that resource manager filtered rules?

**Review by oliver-goetz — COMMENTED** [review]
I added some small comments and have two general questions where I don't know enough about the context to answer them by myself 😄 

When I see a shoot starting up `etcd-main` and `etcd-events` are starting at the same time on different nodes.
a) When there is no zone pinning annotation yet, couldn't it happen, that these etcd nodes would be in located in different zones? If this could happen, their volumes would be in different zones and couldn't be pinned to the same zone anymore right?
b) How about the migration scenario when introducing zone pinning. There are other control-plane components like loki and prometheus which use PVCs. Could it be possible that their volumes are located in different zones and could not be attached anymore when zone pinning was activated?

File: pkg/gardenlet/controller/shoot/shoot_control_reconcile.go (hunk 2 above)
Comment (timuthy) [reply pkg/gardenlet/controller/shoot/shoot_control_reconcile.go:288]: They won't be added to another zone because the pod affinity make sure that the pods are scheduled to the same AZ.
```
   podAffinity:
      requiredDuringSchedulingIgnoredDuringExecution:
      - LabelSelector: {}
        TopologyKey: topology.kubernetes.io/zone
```
The zone information is only required if the entire control-plane is scheduled again (point in time when no pods are scheduled), i.e. that the first scheduled pod lands in the same AZ is was scheduled before. This usually happens when a cluster is woken up from hibernation.

File: pkg/resourcemanager/webhook/podzoneaffinity/handler.go (hunk 3 above)
Comment (timuthy) [reply pkg/resourcemanager/webhook/podzoneaffinity/handler.go]: I have never considered that option. It's surely good for debugging and troubleshooting. A log entry could also do the job. Any opinions @rfranzke?

File: pkg/apis/core/v1beta1/constants/types_constants.go [hunk 5]
```diff
@@ -256,6 +256,9 @@ const (
 	// GardenRoleExposureClassHandler is the value of the GardenRole key indicating type 'exposureclass-handler'.
 	GardenRoleExposureClassHandler = "exposureclass-handler"
 
+	// ShootZonePinning is an annotation key which is used to pin or schedule all control-plane pods
+	// to the very same availability zone.
+	ShootZonePinning = "shoot.gardener.cloud/zone-pinning"
```
Comment (rfranzke) [inline pkg/apis/core/v1beta1/constants/types_constants.go]: Can we use a more expressive name, something like `control-plane.shoot.gardener.cloud/enforce-zone` or similar? Without much context, `zone-pinning` wasn't quite clear to me in the first place.

File: pkg/operation/botanist/namespaces.go (hunk 1 above)
Comment (rfranzke) [reply pkg/operation/botanist/namespaces.go]: That's probably out of scope of this PR (migration non-HA to HA shoots) since it requires more complexity since the PVCs of etcd are always bound to a single zone.

However, if the feature gate gets disabled, the label should probably be removed again from the namespace?

File: pkg/gardenlet/controller/shoot/shoot_control_reconcile.go (hunk 2 above)
Comment (rfranzke) [reply pkg/gardenlet/controller/shoot/shoot_control_reconcile.go:288]: What if the GRM's cache injecting the `podAffinity` via the new webhook is stale?

File: pkg/operation/botanist/namespaces.go [hunk 6]
```diff
@@ -78,15 +79,81 @@ func (b *Botanist) DeploySeedNamespace(ctx context.Context) error {
...
+	}
+
+	namespace := &corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: b.Shoot.SeedNamespace,
+		},
+	}
+
+	if _, err := controllerutils.GetAndCreateOrMergePatch(ctx, b.K8sSeedClient.Client(), namespace, func() error {
+		metav1.SetMetaDataLabel(&namespace.ObjectMeta, v1beta1constants.ShootZonePinning, zone)
```
Comment (rfranzke) [inline pkg/operation/botanist/namespaces.go]: Use https://github.com/gardener/gardener/blob/9044fe3cd5c7c59871bc4de9aca914686f0dc4d3/pkg/operation/types.go#L78

Do not call `GetAndCreateOrMergePatch` but rather `Patch` instead (the namespace should already exist).

File: docs/concepts/resource-manager.md [hunk 7]
```diff
@@ -519,3 +519,37 @@ Hence, it is served from the Seed GRM and each Shoot GRM.
 Please find an overview below for pods deployed in the Shoot cluster:
 
 ![image](images/resource-manager-projected-token-shoot-to-shoot-apiserver.jpg)
+
+### Pod Zone Affinity
+
+When this webhook is activated and namespaces are annotated with `shoot.gardener.cloud/zone-pinning` then it automatically adds a [pod affinity](https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#inter-pod-affinity-and-anti-affinity):
```
Comment (rfranzke) [inline docs/concepts/resource-manager.md]:
```suggestion
When this webhook is activated and namespaces are annotated with `shoot.gardener.cloud/zone-pinning` then it automatically adds a [pod affinity](https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#inter-pod-affinity-and-anti-affinity) to all `Pod`s created in these namespaces:
```

File: docs/concepts/resource-manager.md [hunk 8]
```diff
@@ -519,3 +519,37 @@ Hence, it is served from the Seed GRM and each Shoot GRM.
...
+
+```
+spec:
+  affinity:
+    podAffinity:
+      requiredDuringSchedulingIgnoredDuringExecution:
+      - LabelSelector: {}
+        TopologyKey: topology.kubernetes.io/zone
```
Comment (rfranzke) [inline docs/concepts/resource-manager.md]:
```suggestion
      - labelSelector: {}
        topologyKey: topology.kubernetes.io/zone
```

File: pkg/operation/seed/components.go [hunk 9]
```diff
@@ -149,6 +149,7 @@ func defaultGardenerResourceManager(c client.Client, seedClientVersion string, i
 			},
 		},
 		DefaultSeccompProfileEnabled: gardenletfeatures.FeatureGate.Enabled(features.DefaultSeccompProfile),
+		PodZoneAffinityEnabled:       true,
```
Comment (rfranzke) [inline pkg/operation/seed/components.go]: Shouldn't we only enable this if the gardenlet's feature gate is enabled?

File: pkg/resourcemanager/webhook/podschedulername/podschedulername_suite_test.go [hunk 10]
```diff
@@ -23,5 +23,5 @@ import (
 
 func TestPodSchedulerName(t *testing.T) {
 	RegisterFailHandler(Fail)
-	RunSpecs(t, "Webhook PodSchedulerName Suite")
+	RunSpecs(t, "Webhook PodZoneAffinity Suite")
```
Comment (rfranzke) [inline pkg/resourcemanager/webhook/podschedulername/podschedulername_suite_test.go]: Why was this changed?

File: pkg/resourcemanager/webhook/podzoneaffinity/handler.go [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+	h.client = cl
+	return nil
+}
+
+func (h *handler) InjectDecoder(d *admission.Decoder) error {
+	h.decoder = d
+	return nil
+}
+
+func (h *handler) Handle(ctx context.Context, req admission.Request) admission.Response {
```
Comment (rfranzke) [inline pkg/resourcemanager/webhook/podzoneaffinity/handler.go:59]: Please add an integration test to https://github.com/gardener/gardener/tree/master/test/integration/resourcemanager

File: pkg/resourcemanager/webhook/podzoneaffinity/handler.go (hunk 3 above)
Comment (rfranzke) [reply pkg/resourcemanager/webhook/podzoneaffinity/handler.go]: A log entry would be good, yes. I would not add an annotation. However, the documentation already states that this might happen, so it shouldn't be surprising to operators (all this is only happening in seeds).

File: pkg/resourcemanager/webhook/podzoneaffinity/handler.go [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+			zoneTermExisting = true
+		}
+
+		// If there is another affinity configured on zones, we assume that this will be conflicting.
+		if term.TopologyKey == corev1.LabelTopologyZone && !zoneTermExisting {
+			continue
+		}
+
+		filteredAffinityTerms = append(filteredAffinityTerms, term)
+	}
```
Comment (rfranzke) [inline pkg/resourcemanager/webhook/podzoneaffinity/handler.go]: Is there the option to harmonize this code with (more or less the same) code above by introducing a function operating on `[]PodAffinityTerm`?

File: pkg/resourcemanager/webhook/podzoneaffinity/handler.go [hunk 13]
```diff
@@ -0,0 +1,224 @@
...
+func getZoneSpecificNodeSelector(ctx context.Context, cl client.Client, namespace string) (*corev1.NodeSelectorTerm, error) {
+	namespaceObj := &corev1.Namespace{}
+	if err := cl.Get(ctx, kutil.Key(namespace), namespaceObj); err != nil {
+		return nil, err
+	}
+
+	// Check if scheduling to a specific zone is required.
+	var nodeSelector *corev1.NodeSelectorTerm
+	zone := namespaceObj.Labels[gardencorev1beta1constants.ShootZonePinning]
+	if zone != "" {
```
Comment (rfranzke) [inline pkg/resourcemanager/webhook/podzoneaffinity/handler.go]:
```suggestion
	if zone := namespaceObj.Labels[gardencorev1beta1constants.ShootZonePinning]; zone != "" {
```

**Review by rfranzke — CHANGES_REQUESTED** [review]

File: pkg/gardenlet/controller/shoot/shoot_control_reconcile.go (hunk 2 above)
Comment (timuthy) [reply pkg/gardenlet/controller/shoot/shoot_control_reconcile.go:288]: Do you have any specific scenario in mind? I can't imagine how the `deployControlPlane` task dependency mentioned by Oliver can mitigate this problem.
Adding the `nodeAffinity` (with concrete zone) isn't time critical as it safeguards only the scenario when no control-plane pods run after the shoot was created.
Let's say:
1. Shoot is being created.
2. Namespace is labelled `zone-pinning`, no value.
3. Etcd pod(s) are scheduled with common `podAffinity` injected by GRM (first time a volume is created and thus important to figure out zone being used).
4. Add zone value to `zone-pinning` label.
5. Some more pods are scheduled at the same time: A handful will get only the common `podAffinity`, some other will also get the `nodeAffinity` containing the zone information. The latter is actually not needed at this moment, since the `podAffinity` takes care that pods are still scheduled to the same zone (as long as there is at least one pod running in the shoot namespace in the seed).
6. Creation succeeds.
7. Cluster is hibernated.
8. Cluster is being woken up -> no pods are running in shoot namespace (this situation can of course also happen for other reasons, e.g. seed outage).
9. Etcd pod(s) are scheduled: Now it's important that the original zone from 3. is used. This would only be the case if there is no other pod in the seed namespace at that time. However if there was another pod scheduled to a different zone, then the `etcd` pod(s) would be stuck in `Pending` because of the `podAffinity` and the volume dependency. Hence, with the `nodeAffinity` we make sure that all the pods are scheduled to the original zone, it serves as a safety measure. It's also the earliest point in time we actually need the `nodeAffinity`.

**Comment by timuthy** [comment]
@oliver-goetz thanks a lot for your feedback:

Concerning your questions: It's assumed that there are no seed clusters with worker nodes across multiple zones today that host "normal" shoot clusters. As discussed, I'll add a release note explaining the breaking change.

**Comment by timuthy** [comment]
Thanks for your feedback @rfranzke and @oliver-goetz. I addressed all your comments, open is still https://github.com/gardener/gardener/pull/6579#discussion_r962570312. PTAL.

File: pkg/gardenlet/controller/shoot/shoot_control_reconcile.go (hunk 2 above)
Comment (rfranzke) [reply pkg/gardenlet/controller/shoot/shoot_control_reconcile.go:288]: OK, thanks for thinking it through. I don't have any specific example, just wanted to raise this question since we had interesting behaviours in the past when caches were stale :)

**Comment by rfranzke** [comment]
> ```
> Gardener has been being prepared for more shoot HA use-cases and thus some assumption about currently running landscapes are required: If you use a `multi-zonal` labelled seed and scheduled non-HA shoots onto it, this release of Gardener will potentially cause scheduling conflicts to the control-plane pods as it will try to locate all pods into a single zone only. Pods that can't be re-scheduled (mainly because of volume dependencies) will remain in `Pending` state.
> ```

Just to double-check: This is likely no problem because the `"seed.gardener.cloud/multi-zonal"` label must be added manually to the `Seed`s, right? I can imagine that there are folks in the wild that have seeds with multiple AZs, however I believe that nobody has added this label yet given that it's quite new and the whole HA story is still alpha/in development. In this case, I'm fine with it.

File: pkg/operation/botanist/namespaces.go [hunk 14]
```diff
@@ -146,16 +147,10 @@ func (b *Botanist) AddZoneInformationToSeedNamespace(ctx context.Context) error
...
-		},
-	}
+	namespace := b.SeedNamespaceObject.DeepCopy()
+	metav1.SetMetaDataLabel(&namespace.ObjectMeta, v1beta1constants.ShootControlPlaneEnforceZone, zone)
 
-	if _, err := controllerutils.GetAndCreateOrMergePatch(ctx, b.K8sSeedClient.Client(), namespace, func() error {
-		metav1.SetMetaDataLabel(&namespace.ObjectMeta, v1beta1constants.ShootZonePinning, zone)
-		return nil
-	}); err != nil {
+	if err := b.K8sSeedClient.Client().Patch(ctx, namespace, client.MergeFrom(b.SeedNamespaceObject)); err != nil {
```
Comment (rfranzke) [inline pkg/operation/botanist/namespaces.go]: Nit: Typically, we follow this pattern:

```golang
patch := client.MergeFrom(object.DeepCopy())
object.Spec.Foo = "bar" // whatever modifications on the object
if err := client.Patch(ctx, object, patch); err != nil {
   ...
}
```

File: test/integration/resourcemanager/podzoneaffinity/podzoneaffinity_test.go [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+			Spec: corev1.PodSpec{
+				Containers: []corev1.Container{
+					{
+						Name:  "foo-container",
+						Image: "foo",
+					},
+				},
+			},
+		}
+		DeferCleanup(func() {
```
Comment (rfranzke) [inline test/integration/resourcemanager/podzoneaffinity/podzoneaffinity_test.go:45]: Nit for readability:

```suggestion

		DeferCleanup(func() {
```

File: test/integration/resourcemanager/podzoneaffinity/podzoneaffinity_test.go [hunk 16]
```diff
@@ -0,0 +1,133 @@
...
+			Expect(testClient.Delete(ctx, pod)).To(Succeed())
+		})
+	})
+
+	Context("when namespace has zone enforcement label", func() {
+		BeforeEach(func() {
+			testNamespace.Labels = map[string]string{
+				"control-plane.shoot.gardener.cloud/enforce-zone": "",
+			}
+			Expect(testClient.Update(ctx, testNamespace)).To(Succeed())
```
Comment (rfranzke) [inline test/integration/resourcemanager/podzoneaffinity/podzoneaffinity_test.go]: Are there situations where these `Update` calls can lead to conflicts (maybe during stress-testing)? If yes, you could switch to `Patch`.

File: test/integration/resourcemanager/podzoneaffinity/podzoneaffinity_test.go [hunk 17]
```diff
@@ -0,0 +1,133 @@
...
+			DeferCleanup(func() {
+				testNamespace.Labels = nil
+				Expect(testClient.Update(ctx, testNamespace)).To(Succeed())
+			})
+		})
+
+		It("should add podAffinity", func() {
+			Expect(testClient.Create(ctx, pod)).To(Succeed())
+
+			Consistently(func() ([]corev1.PodAffinityTerm, error) {
```
Comment (rfranzke) [inline test/integration/resourcemanager/podzoneaffinity/podzoneaffinity_test.go]: Why would we have to use async assertions in such webhook tests? I see that you probably copied this from other existing tests, but shouldn't the `pod` already have been modified right during creation (otherwise, the above `Create` call failed)?
If you just copied this and there is no reason for it, please remove it (maybe also cleanup the other test suites as you are on it) 👀

File: test/integration/resourcemanager/podzoneaffinity/podzoneaffinity_test.go [hunk 18]
```diff
@@ -0,0 +1,133 @@
...
+				Expect(testClient.Get(ctx, client.ObjectKeyFromObject(pod), pod)).To(Succeed())
```
Comment (rfranzke) [inline test/integration/resourcemanager/podzoneaffinity/podzoneaffinity_test.go]: If async assertions are needed/kept: You should pass `g Gomega` to your `func` and use `g.Expect` here.
(same below if needed)

**Review by rfranzke — COMMENTED** [review]
Thanks for addressing my comments @timuthy!

**Comment by timuthy** [comment]
@rfranzke 

> Just to double-check: This is likely no problem because the "seed.gardener.cloud/multi-zonal" label must be added manually to the Seeds, right?

Yes, that's mostly right. We planned to enable automatic labeling for ManagedSeeds because there we have information about workers.

**Comment by timuthy** [comment]
Thanks @rfranzke. I incorporated your feedback and in addition found some spots where I could improve the test handling regarding your suggestions: [56a86cf](https://github.com/gardener/gardener/pull/6579/commits/56a86cf9be727c1a12f93c96cf9b22b3d3ce8505), [b4a4869](https://github.com/gardener/gardener/pull/6579/commits/b4a486943349bcd4f24eda8a53ee24b6ec2eaf73)

**Review by rfranzke — COMMENTED** [review]
Cool @timuthy, thanks for also cleaning up the other webhook test suites!

**Comment by rfranzke** [comment]
/approve

@oliver-goetz Do you also want to have another look?

File: pkg/operation/botanist/namespaces.go [hunk 19]
```diff
@@ -79,8 +79,9 @@ func (b *Botanist) DeploySeedNamespace(ctx context.Context) error {
 
 		// Label namespace to pin all control-plane pods of a shoot cluster to one zone
 		// if the seed has workers across different availability zones.
-		if zonePinningRequired(b.Shoot.GetInfo(), b.Seed.GetInfo()) && !metav1.HasLabel(namespace.ObjectMeta, v1beta1constants.ShootZonePinning) {
-			metav1.SetMetaDataLabel(&namespace.ObjectMeta, v1beta1constants.ShootZonePinning, "")
+		delete(namespace.Labels, v1beta1constants.ShootControlPlaneEnforceZone)
+		if zonePinningRequired(b.Shoot.GetInfo(), b.Seed.GetInfo()) && !metav1.HasLabel(namespace.ObjectMeta, v1beta1constants.ShootControlPlaneEnforceZone) {
+			metav1.SetMetaDataLabel(&namespace.ObjectMeta, v1beta1constants.ShootControlPlaneEnforceZone, "")
 		}
```
Comment (oliver-goetz) [inline pkg/operation/botanist/namespaces.go]:
```suggestion
		if !zonePinningRequired(b.Shoot.GetInfo(), b.Seed.GetInfo()) {
			delete(namespace.Labels, v1beta1constants.ShootControlPlaneEnforceZone)
		} else if !metav1.HasLabel(namespace.ObjectMeta, v1beta1constants.ShootControlPlaneEnforceZone) {
			metav1.SetMetaDataLabel(&namespace.ObjectMeta, v1beta1constants.ShootControlPlaneEnforceZone, "")
		}
```

Otherwise, the label would be changed back and forth during reconciliation, if enforce zone is activated.
This function would set its value to `""`, but `AddZoneInformationToSeedNamespace` would add a zone value again later.

**Review by oliver-goetz — CHANGES_REQUESTED** [review]
There is one small thing left.

File: pkg/operation/botanist/namespaces.go (hunk 19 above)
Comment (timuthy) [reply pkg/operation/botanist/namespaces.go]: Nice catch @oliver-goetz. I fixed it, PTAL.

**Review by oliver-goetz — APPROVED** [review]
/lgtm

🚀
